uikit/poidetail/src/main/java/com/sygic/ui/view/poidetail/viewmodel/PoiDetailInternalViewModel.kt

```diff
@@ -0,0 +1,106 @@
+package com.sygic.ui.view.poidetail.viewmodel
+
+import androidx.lifecycle.*
+import com.google.android.material.bottomsheet.BottomSheetBehavior
+import com.sygic.ui.common.behaviors.BottomSheetBehaviorWrapper
+import com.sygic.ui.common.livedata.SingleLiveEvent
+import com.sygic.ui.common.sdk.data.PoiData
+import com.sygic.ui.common.sdk.extension.getFormattedLocation
+import com.sygic.ui.view.poidetail.PoiDetailBottomDialogFragment
+import com.sygic.ui.view.poidetail.manager.PreferencesManager
+import kotlinx.coroutines.*
+import java.util.concurrent.TimeUnit
+
+private val SHOWCASE_DELAY = TimeUnit.SECONDS.toMillis(1)
+const val DEFAULT_BEHAVIOR_STATE = BottomSheetBehavior.STATE_COLLAPSED
+const val SHOWCASE_BEHAVIOR_STATE = BottomSheetBehavior.STATE_EXPANDED
+
+internal class PoiDetailInternalViewModel(poiData: PoiData,
+                                 private val preferencesManager: PreferencesManager) : ViewModel(),
+    BottomSheetBehaviorWrapper.StateListener {
+
+    val titleText: String
+    val subtitleText: String
+    val coordinatesText: String?
+    val urlText: String? = poiData.url
+    val emailText: String? = poiData.email
+    val phoneText: String? = poiData.phone
+
+    val expandObservable: LiveData<Any> = SingleLiveEvent()
+    val collapseObservable: LiveData<Any> = SingleLiveEvent()
+    val webUrlClickObservable: LiveData<String> = SingleLiveEvent()
+    val emailClickObservable: LiveData<String> = SingleLiveEvent()
+    val phoneNumberClickObservable: LiveData<String> = SingleLiveEvent()
+    val coordinatesClickObservable: LiveData<String> = SingleLiveEvent()
+
+    private var listener: PoiDetailBottomDialogFragment.Listener? = null
+    private var showcaseLaunch: Job? = null
+
+    init {
+        val addressComponent = poiData.getAddressComponent()
+        titleText = addressComponent.formattedTitle
+        subtitleText = addressComponent.formattedSubtitle
+        coordinatesText = poiData.coordinates.getFormattedLocation()
+    }
+
+    fun onHeaderClick() {
+        (expandObservable as SingleLiveEvent<Any>).call()
+    }
+
+    fun onWebUrlClick() {
+        (webUrlClickObservable as SingleLiveEvent<String>).value = urlText
```

Comment:
rozmyslam, ze by sme mohli vytvorit takuto extension funkciu

```fun <T : Any> LiveData<T>.asMutable(): MutableLiveData<T> {
    return if (this is MutableLiveData<T>) this else throw IllegalArgumentException("$this is not an instance of MutableLiveData")
}```

a potom by sme mohli nemuseli castovat takto, ale iba volanim `poiDataObservable.asMutable().value = poiData`. Pre single by mohlo byt tiez vlastna metoda. Ani tento zapis nie je zly, je to vidiet, co sa castuje, ale kedze sa to casto opakuje, zapisovanie nie je najprijemnejsie :/

Reply:
jj, paráda, určitě použijeme, díky za tip ;)